packages/astro/src/runtime/client/dev-overlay/plugins/audit/a11y.ts

```diff
@@ -280,14 +282,20 @@ export const a11y: AuditRuleWithSelector[] = [
 		selector: 'a[href]:is([href=""], [href="#"], [href^="javascript:" i])',
 	},
 	{
-		code: 'a11y-label-has-associated-control',
-		title: '`label` tag should have an associated control and a text content.',
+		code: 'a11y-invalid-label',
+		title: '`label` element should have an associated control and a text content.',
 		message:
-			'The `label` tag must be associated with a control using either `for` or having a nested input. Additionally, the `label` tag must have text content.',
-		selector: 'label:not([for])',
+			'The `label` element must be associated with a control using either `for` or having a nested form element. Additionally, the `label` element must have text content.',
+		selector: 'label',
 		match(element) {
-			const inputChild = element.querySelector('input');
-			if (!inputChild?.textContent) return true;
+			// Label must be associated with a control
+			const hasFor = element.hasAttribute('for');
+			const nestedFormElement = element.querySelector(`${formElements.join(', ')}`);
+			if (!hasFor && !nestedFormElement) return true;
+
+			// Label must have text content
+			const innerText = (element as HTMLLabelElement).innerText.trim();
+			if (innerText === undefined || innerText.length === 0) return true;
```

Comment:
This is not perfect, because some form elements might have text that are not actually labels and the audit can't tell that that label element is then potentially unnecessary (though there's no actual a11y issue, I think)

It's still an improvement, though!

Reply:
The check is fine with me, but the logic seems wrong? If we're doing `.trim()` we already have a string value ad don't need to check for `undefined`.

```suggestion
			// Label must have text content
			const innerText = (element as HTMLLabelElement).innerText.trim();
			if (innerText === '') return true;
```

Reply:
I'm also a bit confused if `innerText` was an intentional choice vs the standard `textContent`. https://kellegous.com/j/2013/02/27/innertext-vs-textcontent/

`innerText` does take layout into account, so I can see how that would be relevant here, but just wanted to double check.

Reply:
Yes, `innerText` is an intentional choice, I figured that for accessibility it was better since it takes into account CSS and stuff

Reply:
Cool, that's valid. A comment to that effect would be helpful!

My original suggestion about the conditional logic still stands. `innerText` (seemingly) won't ever be `undefined`.